### PR TITLE
Un-Vendor the net-ssh library

### PR DESCRIPTION
**Pull request by thelightcosine**
MS-1688

This PR removes the vendored version of net-ssh and uses the real gem as a standard dependency.
It add the SSHFactory class as a way to abusing Net::SSH proxy behaviour to substitute our own Rex::sockets for standard ones inside the transport layer.

We will need to verify as many of the SSH based modules as possible, especially ssh_login and ssh_login_pubkey (and ensuring ssh sessions from those modules still work)

### CODE REVIEW

File: modules/auxiliary/ssh_test.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+##
```
Comment (wvu-r7) [inline modules/auxiliary/ssh_test.rb]: Wrong place for this file?

File: lib/rex/socket/ssh_factory.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  module Socket
+
+    # This class exists to abuse the Proxy capabilities in the Net::SSH library to allow the use of Rex::Sockets
+    # for the transport layer in Net::SSH. The SSHFactory object will respond to the #open method and create the
+    # {Rex::Socket::Tcp}
+    class SSHFactory
+
+      # @!attribute msfraemwork
+      #   @return [Object] The framework instance object
+      attr_accessor :msframework
```
Comment (egypt) [inline lib/rex/socket/ssh_factory.rb]: Everywhere else in the code, a framework object is just called `framework`. It seems weird to buck that convention.

**Comment by sempervictus** [comment]
Pondering this over, and i'm wondering what the long-term plan for Socket vs Proto is in these cases. SSH and SSL both wrap sockets in their own way, but contain functionality which isn't directly socket related. Maybe we start thinking of SSH as a Proto instead of a Socket type, and abstract the protocol forwarding as modules in the Proto section, and subsequently deriving a Transport from the modules.

Once this tear-down is complete, i'd like to take a stab at the low-level concurrency and socket segments to implement LibUV sockets, event loops, and contracts on the lower tiers to reduce our threadcount and provide a consistent event handling implementation for monitors, callbacks, etc. This should also allow us to move pivot sockets and other virtual abstractions here. The more things we have in Socket and Sync (or whatever sync will become), the more cumbersome that task becomes.

Thoughts?

File: modules/auxiliary/gather/apache_karaf_command_execution.rb [hunk 3]
```diff
@@ -71,16 +71,17 @@ def cmd
   end
 
   def do_login(user, pass, ip)
+    factory = Rex::Socket::SSHFactory.new(framework,self, datastore['Proxies'])
```
Comment (sempervictus) [inline modules/auxiliary/gather/apache_karaf_command_execution.rb]: Any chance we create an Msf::Exploit::Remote::SSH namespace for this and other convenience methods?

File: lib/rex/socket/ssh_factory.rb (hunk 2 above)
Comment (dmaloney-r7) [inline lib/rex/socket/ssh_factory.rb]: It was copied over from how we passed it into net-ssh, i can change it though

File: modules/auxiliary/gather/apache_karaf_command_execution.rb (hunk 3 above)
Comment (dmaloney-r7) [inline modules/auxiliary/gather/apache_karaf_command_execution.rb]: yes, that should actually get refactored so we don't have copy-paste everywhere

File: lib/msf/core/exploit/ssh.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+module Msf
+  module Exploit
+    module Remote
+      module SSH
+
+        def ssh_socket_factory
+          Rex::Socket::SSHFactory.new(framework,self, datastore['Proxies'])
```
Comment (sempervictus) [inline lib/msf/core/exploit/ssh.rb]: Any reason not to reuse a factory after instantiation?

**Comment by sempervictus** [comment]
@dmaloney-r7: while you're carving out net-ssh from Framework, have you seen any changes this community has made to the library that we could push back to the official repo? 
Everything i've done has been MSF-transport or info-leak specific. The partial server implementation can be moved over to the official branch and completed there, the proxies and sockets pieces dont need to go up.

**Comment by dmaloney-r7** [comment]
@sempervictus pretty much everything we've done has been either transport(Rex::Socket) related or back porting fixes from the upstream gem

**Comment by sempervictus** [comment]
@dmaloney-r7: thanks for the clarification. I recall the net-ssh gem maintenance having gone stale and thinking at one point we were actually ahead of them in some way. Just hoping we dont accidentally lose some useful bits along the way

**Comment by bcook-r7** [comment]
I believe there is another change in that we disabled net::ssh's tendency to look at ~/.ssh/ for configuration of its default behavior.

**Comment by sempervictus** [comment]
Thats a good point, i had to do something similar with DNS recently to prevent the resolver from hitting user-default NS'. We should probably review all the vendor libraries for use of local configuration files which could lead to info leaks.

**Comment by dmaloney-r7** [comment]
This was already handled by the upstream Gem a long time ago, in fact the change we made in this regards was actually back porting the fix from the original maintainer. This functionality already exists

**Comment by dmaloney-r7** [comment]
I have audited all of the commits that we made on our copy of net-ssh during this process to ensure we didn't lose anything vital.

**Comment by sempervictus** [comment]
Thank you very much, and apologies for complicating what's already a
massive undertaking.

**Comment by dmaloney-r7** [comment]
no worries, we need lots of eyes on things like this to make sure nothing gets missed

**Comment by egypt** [comment]
When the password is wrong, this is prompting on `$stdout`:
```
14:49:13 netssh j:0 s:1 auxiliary(ssh_login) > set PASSWORD notpassword
PASSWORD => notpassword
14:49:18 netssh j:0 s:1 auxiliary(ssh_login) > run

[*] SSH - Starting bruteforce
Text will be echoed in the clear. Please install the HighLine or Termios libraries to suppress echoed text.
msfadmin@192.168.99.2's password:
```

Seems to be working correctly when the password is correct:

```
14:49:11 netssh j:0 s:0 auxiliary(ssh_login) > run

[*] SSH - Starting bruteforce
[+] SSH - Success: 'msfadmin:msfadmin' 'uid=1000(msfadmin) gid=1000(msfadmin) groups=4(adm),20(dialout),24(cdrom),25(floppy),29(audio),30(dip),44(video),46(plugdev),107(fuse),111(lpadmin),112(admin),119(sambashare),1000(msfadmin) Linux metasploitable 2.6.24-16-server #1 SMP Thu Apr 10 13:58:00 UTC 2008 i686 GNU/Linux '
[*] Command shell session 1 opened (192.168.99.1:45259 -> 192.168.99.2:22) at 2016-07-13 14:49:13 -0500
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

**Comment by dmaloney-r7** [comment]
Net::SSH added a non_interactive flag for this, it just needed to be set, should be good now

File: modules/exploits/apple_ios/ssh/cydia_default_ssh.rb [hunk 5]
```diff
@@ -72,16 +72,18 @@ def rport
...
   def do_login(user, pass)
+    factory = ssh_socket_factory
     opts = {
       :auth_methods => ['password', 'keyboard-interactive'],
-      :msframework  => framework,
+      :framework  => framework,
       :msfmodule    => self,
       :port         => rport,
-      :disable_agent => true,
+      :use_agent => true,
```
Comment (egypt) [inline modules/exploits/apple_ios/ssh/cydia_default_ssh.rb]: should be `false`

File: modules/exploits/linux/ssh/f5_bigip_known_privkey.rb [hunk 6]
```diff
@@ -70,16 +70,18 @@ def rport
...
   def do_login(user)
+    factory = Rex::Socket::SSHFactory.new(framework,self, datastore['Proxies'])
     opt_hash = {
       :auth_methods => ['publickey'],
       :msframework  => framework,
       :msfmodule    => self,
       :port         => rport,
       :key_data     => [ key_data ],
-      :disable_agent => true,
+      :use_agent => true,
```
Comment (egypt) [inline modules/exploits/linux/ssh/f5_bigip_known_privkey.rb]: should be `false`

File: modules/exploits/multi/ssh/sshexec.rb [hunk 7]
```diff
@@ -100,14 +101,15 @@ def execute_command(cmd, opts = {})
...
 
   def do_login(ip, user, pass, port)
+    factory = ssh_socket_factory
     opt_hash = {
       :auth_methods  => ['password', 'keyboard-interactive'],
-      :msframework   => framework,
-      :msfmodule     => self,
       :port          => port,
-      :disable_agent => true,
+      :use_agent     => true,
```
Comment (egypt) [inline modules/exploits/multi/ssh/sshexec.rb]: should be `false`

**Comment by wchen-r7** [comment]
# Release Notes

No release notes needed.

**Pull request merged**